cyclic_boosting/quantile_matching.py

```diff
@@ -161,6 +166,571 @@ def cdf(self, x):
+    alpha : float
+        lower quantile of SPT (upper is ``1 - alpha``)
+    l : float
+        lower bound of supported range (only active for bound and semi-bound
+        modes)
+    u : float
+        upper bound of supported range (only active for bound mode)
+    """
+
+    def __init__(
+        self,
+        est_median: BaseEstimator,
+        est_lowq: BaseEstimator,
+        est_highq: BaseEstimator,
+        bound: str,
+        alpha: Optional[float] = 0.25,
+        l: Optional[float] = 0,
+        u: Optional[float] = 1,
+    ):
+        self.est_median = est_median
+        self.est_lowq = est_lowq
+        self.est_highq = est_highq  # trained with quantile 2 * (1 - alpha) - 1 (e.g., median for 0.75)
+
+        self.alpha = alpha
+
+        self.l = l
+        self.u = u
+
+        self.bound = bound
+        if self.bound not in ["S", "B", "U"]:
+            raise Exception("Invalid version.")
+
+        self.shape = 0
+
+    def fit(self, X: Union[pd.DataFrame, np.ndarray], y: np.ndarray) -> BaseEstimator:
+        # fit sinhlogistic shape parameter
+        self.shape = fit_sinhlogistic_shape(self.alpha, self.l, self.u, self.bound, y)
+
+        # median model
+        if self.bound == "S":
+            y_trans = transform_from_semibound_lower(y, self.l)
+        elif self.bound == "B":
+            y_trans = transform_from_bounds(y, self.l, self.u)
+        else:
+            y_trans = y
+        self.est_median.fit(X, y_trans)
+        z = self.est_median.predict(X)
+        if self.bound == "S":
+            pred_median = back_transform_in_semibound_lower(z, self.l)
+        elif self.bound == "B":
+            pred_median = back_transform_in_bounds(z, self.l, self.u)
+        else:
+            pred_median = z
+
+        # lower quantile model
+        y_trans = y / pred_median
+        mask = y_trans < 1
+        y_trans = y_trans[mask]
+        if self.bound in ["S", "B"]:
+            y_trans = transform_from_bounds(y_trans, self.l / pred_median[mask], 1)
+        else:
+            y_trans = transform_from_semibound_upper(y_trans, 1)
+        self.est_lowq.fit(X[mask], y_trans)
+        z = self.est_lowq.predict(X)
+        if self.bound in ["S", "B"]:
+            pred_lowq = back_transform_in_bounds(z, self.l, pred_median)
+        else:
+            pred_lowq = back_transform_in_semibound_upper(z, pred_median)
+
+        # upper quantile model
+        y_trans = (y - pred_median) / (pred_median - pred_lowq)
+        mask = y_trans > 0
+        y_trans = y_trans[mask]
+        if self.bound in ["S", "U"]:
+            y_trans = transform_from_semibound_lower(y_trans, 0)
+        else:
+            y_trans = transform_from_bounds(y_trans, 0, (self.u - pred_median) / (pred_median - pred_lowq))
+        self.est_highq.fit(X[mask], y_trans)
+
+        return self
+
+    def predict(self, X: Union[pd.DataFrame, np.ndarray]) -> np.ndarray:
+        if self.bound == "S":
```

Comment:
I do not have a good suggestion ATM, but this could likely be rewritten so that `pred_median, pred_lowq, pred_highq` are returned as a tuple from a method of a `BackTransformation` class, containing the `back_transform_` functions above.

Reply:
Hm, I don't see a real advantage in putting these functions in a class.
What might be a nicer implementation of the chain is a pipeline. But I decided against it because the cuts make it complicated.

Reply:
Good point.